### DISCUSSION TITLE
How to use custom TMS/projection with 'Mosaic from COG urls' example

### DISCUSSION THREAD

**Discussion by Shane98c** [developmentseed/titiler#591] [Q&A] [answered]
Thanks for such a powerful tool!

I've got the example from here working: https://developmentseed.org/titiler/examples/code/mosaic_from_urls/

It seems fast and great!

My next requirement is to get this working for polar projections. I've been able to add a custom tms to the route like so:

```
EPSG3031 = TileMatrixSet.custom(
    (-948.75, -543592.47, 5817.41, -3333128.95),
    CRS.from_epsg(3031),
    identifier="EPSG3031",
)

tms = tms.register([EPSG3031])

...

mosaic = MosaicTiler(reader=MultiFilesBackend, path_dependency=DatasetPathParams, supported_tms=tms, default_tms="EPSG3031")
``` 
It shows up under /docs and appears to be generally working. However, the output is not changed, and tiles are still returned in web mercator when I hit the endpoint with a request like this: 
`http://127.0.0.1:8000/tiles/EPSG3031/7/44/93@4x?url=https://pgc-opendata-dems.s3.us-west-2.amazonaws.com/rema/strips/s2s041/2m/s64w056/SETSM_s2s041_WV01_20180922_102001007A2EAF00_102001007D190200_2m_lsf_seg1_dem_10m_shade_masked.tif`
(I would expect this request to fail since the zxy are defined for the location in web mercator tms, instead it just returns the tile)

I think this all has to do with using this: `from cogeo_mosaic.backends.base import BaseBackend` which I've seen in a couple [code comments](https://github.com/developmentseed/cogeo-mosaic/blob/2bf5063564f685ffd64f4ce361432a1814eb659e/cogeo_mosaic/backends/base.py#L63) only supports web mercator.

Is there an example someone could point me to which sets up a backend that can mosaic multiple cogs (provided via list in url) like the example above but also supports projection to custom tms (and accepting tile requests in that custom tms as well)

I feel like I might be missing something fundamental in my mental model here. I really appreciate any help!

**Comment by vincentsarago** [comment, marked as answer]
👋 Hi @Shane98c 

😬 This turns out to be pretty complex. The Mosaic Factory use `BaseBackend` as a reference class. Because in `BaseBackend` the TMS is set with [`init=False`](https://github.com/developmentseed/cogeo-mosaic/blob/2bf5063564f685ffd64f4ce361432a1814eb659e/cogeo_mosaic/backends/base.py#L64) (meaning that user shouldn't pass the tms as input parameter) the MosaicTilerFactory's tiles endpoint do not use TMS as a `reader` input

https://github.com/developmentseed/titiler/blob/418bb938d9cf33df39a3bf25d20f18d23fa64ec3/src/titiler/mosaic/titiler/mosaic/factory.py#L278-L283

This is one of the reason we had to re-write most of the factory for titiler-pgstac https://github.com/stac-utils/titiler-pgstac/blob/master/titiler/pgstac/factory.py#L251-L255

Sadly there is no quick fix (apart re-writing the MosaicTilerFactory `tile` and `tilejson` code on your end.

FYI, I wanted to refactor the multiple file example to switch from mosaic to regular BaseReader but didn't find the time to finish. I'll 👋 you if I do

**Reply by Shane98c** [reply]
Thanks @vincentsarago! Am I correct in understanding that titiler-pgstac has the desired mosaic & projection functionality working already? We're exploring building a dynamic STAC server with pgstac so playing with titiler-pgstac is on my list!

**Reply by vincentsarago** [reply]
> titiler-pgstac has the desired mosaic & projection functionality 

Yes it does

@Shane98c feel free to checkout https://github.com/developmentseed/eoAPI for a `full` pgstac + stac + tiler solution

**Reply by vincentsarago** [reply]
FYI here is a simple mosaic reader you can use with a simple TilerFactory https://gist.github.com/vincentsarago/d2175fb73c62467b116640c1359fc0bd 

Not tested and do not implement all methods

**Reply by Shane98c** [reply]
Thanks so much @vincentsarago! I'll give it a try!
